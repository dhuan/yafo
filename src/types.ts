import React from 'react';

export enum FormFieldType { TEXT, SELECT, RADIO, CHECKBOX }

export type FormFieldComponentProps = {
    inputId        : string
    label          : string
    value          : FormValue
    disabled       : boolean
    onChange       : OnChangeFormValueFunc
    errorMessage   : string
    options        : FormFieldOptions
}

export type FormFieldOption = string | boolean | number | string[]

export type FormFieldOptions = Record<string, FormFieldOption>

export type FieldValidationResult = [ boolean, string ]

export type FieldValidator = <T>(value: FormValue, formValue: (id: T) => FormValue) => FieldValidationResult

export type FormField<T> = {
    id         : T
    label      : string
    type       : FormFieldType
    valid      : FieldValidator
    initial    : FormValue
    options?   : FormFieldOptions
    disabled?  : boolean
}

export type OnChangeFormValueFunc = (newValue: FormValue) => void

export type FormValueFunc = <T>(id: T) => FormValue

export type ValidateFormFunc = () => boolean

export type FormValue = string | number

export type FormProps<T> =  {
    fieldComponents        : Map<T, React.ReactElement>
    dirty                  : boolean
    validate               : ValidateFormFunc
    valid                  : boolean
    value                  : FormValueFunc
    showErrorMessages      : () => void
    hideErrorMessages      : () => void
    disable                : VoidFunction
    enable                 : VoidFunction
    setValues              : (values: Map<T, FormValue>) => void
    setEnterHandler        : (window: Window, handler: VoidFunction) => void
    getInvalidFields       : () => T[]
}

export type FieldComponent = React.FunctionComponent<FormFieldComponentProps>

export type FieldCollection = Map<FormFieldType, FieldComponent>
