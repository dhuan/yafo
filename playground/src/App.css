.app {
    display: flex;
    padding: 10px;
}

.app .form {
    margin-right: 20px;
}

.app .result {
    background: #eee;
    border: 1px solid black;
    padding: 20px;
}
