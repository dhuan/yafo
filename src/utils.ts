export const replaceNth = <T>(list: T[], index: number, newValue: T): T[] =>
    list.map((value: T, i: number) => i === index ? newValue : value);

export const allEqual = <T>(value: T, list: T[]): boolean => {
    const current = list[0];

    if (list.length === 0)
        return false;

    if (current !== value)
        return false;

    if (list.length === 1)
        return true;

    return allEqual(value, list.slice(1));
}

export const unzip = <A, B>(list: [ A, B ][], result: [ A[], B[] ] = [ [], [] ]): [ A[], B[] ] => {
    if (list.length === 0)
        return result;

    const [ a, b ] = list[0];

    const newResult = [ ( result[0] || [] ).concat(a) as A[], ( result[1] || [] ).concat(b) as B[] ];

    return unzip(list.slice(1), newResult as [ A[], B[] ]);
}

export const tupleToMap = <A, B>(tuples: [A, B][], map: Map<A, B>): Map<A, B> => {
    if (tuples.length === 0)
        return map;

    const [ a, b ] = tuples[0];

    map.set(a, b);

    return tupleToMap(tuples.slice(1), map);
}

export const toNumber = (value: any): number => parseInt(value, 10)
