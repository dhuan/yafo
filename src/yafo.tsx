import React, { useState, useEffect } from 'react';
import { replaceNth, allEqual, unzip, mapGetter } from './utils';

import {
    FieldValidationResult,
    FieldValidator,
    Field,
    FieldCollection,
    FieldsDefinition,
    ValueFunc,
    Value,
    Props,
    FieldComponent,
} from './types';

export {
    FieldType,
    FieldComponentProps,
    FieldOption,
    FieldValidationResult,
    FieldValidator,
    Field,
    OnChangeValueFunc,
    FieldCollection,
    FieldsDefinition,
    ValueFunc,
    Value,
    Props,
    FieldComponent,
} from './types';

export { validators } from './validators';

export { fieldCollection } from './field_collection';
export { parseCheckboxFormValue } from './utils';

const changeStateValue = <T extends unknown>(field: Field<T>, state: State, formValue: ValueFunc) => (index: number) => (newValue: Value): void => {
    const { values, setValues, validation, setValidation, dirty, setDirty, errorMessage, setErrorMessage } = state

    const currentValue = values[index]

    const newValues = replaceNth(values, index, newValue)

    const [ valid, fieldErrorMessage ] = field.valid(newValue, formValue)

    const newValidation = replaceNth(validation, index, valid)

    const newErrorMessage = replaceNth(errorMessage, index, fieldErrorMessage)

    const valuesAreEqual = newValue === currentValue

    setValues(newValues)

    setValidation(newValidation)

    setErrorMessage(newErrorMessage)

    if (!dirty && !valuesAreEqual)
        setDirty(true);
}

const getFieldComponent = <T_FieldType extends unknown, T_TargetComponentProps extends unknown>(
    formName            : string,
    field               : Field<T_FieldType>,
    value               : Value,
    changeValue         : (index: number) => (newValue: Value) => void,
    fieldValid          : boolean,
    errorMessageBase    : string,
    index               : number,
    fieldCollection     : FieldCollection,
    formActive          : boolean,
    options             : Options<T_FieldType, T_TargetComponentProps>,
    inputIds            : string[],
): React.ReactElement => {
    const fieldOptions = field.options || []

    const fieldCustomOptions = field.custom || {}

    const errorMessage =
        !fieldValid && options.errorMessagesVisible ? errorMessageBase : "";

    const disabled = !formActive || field.disabled

    const inputId = inputIds[index]

    const Component: FieldComponent = fieldCollection.get(field.type) as FieldComponent

    if (!Component)
        throw new Error(`Could not get form field component for ${field.type}`);

    return (
        <Component
            inputId={inputId}
            label={field.label}
            value={value}
            disabled={disabled || false}
            onChange={changeValue(index)}
            errorMessage={errorMessage}
            options={fieldOptions}
            custom={fieldCustomOptions}
        />
    )
}

const toFieldComponents = <T_FieldType extends unknown, T_TargetComponentProps extends unknown>(
    formName        : string,
    fields          : Field<T_FieldType>[],
    state           : State,
    map             : Map<T_FieldType, React.ReactElement>,
    fieldCollection : FieldCollection,
    options         : Options<T_FieldType, T_TargetComponentProps>,
    formValue       : ValueFunc,
    formActive      : boolean,
    inputIds        : string[],
    index = 0,
): Map<T_FieldType, React.ReactElement> =>
{
    if (index > (fields.length - 1))
        return map;

    const { values, validation, errorMessage } = state;

    const field = fields[index];

    const value = typeof values[index] === "undefined" ? "" : values[index];

    const fieldValid = validation[index]

    const fieldErrorMessage = fieldValid ? "" : errorMessage[index];

    const FieldComponent = getFieldComponent(
        formName,
        field,
        value,
        changeStateValue(field, state, formValue),
        fieldValid,
        fieldErrorMessage,
        index,
        fieldCollection,
        formActive,
        options,
        inputIds,
    );

    map.set(field.id, FieldComponent);

    return toFieldComponents(formName, fields, state, map, fieldCollection, options, formValue, formActive, inputIds, index + 1);
}

type State = {
    values            : Value[]
    validation        : boolean[],
    dirty             : boolean,
    setValues         : (newValues: Value[]) => void,
    setValidation     : (newValidation: boolean[]) => void,
    setDirty          : (newDirty: boolean) => void,
    errorMessage      : string[],
    setErrorMessage   : (newErrorMessage: string[]) => void,
}

const getFormValidation = <T extends unknown>(fields: Field<T>[], state: State, formValue: (id: T) => Value, index = 0): [ boolean, string ][] => {
    if (fields.length === 0)
        return [];

    const [ valid, errorMessage ] = fields[0].valid(state.values[index], formValue);

    return [ [ valid, errorMessage ] as [ boolean, string ] ].concat(getFormValidation(fields.slice(1), state, formValue, index + 1));
}

const validateForm = <T extends unknown>(fields: Field<T>[], state: State, formValue: (id: T) => Value) => (): boolean => {
    const validationResults = getFormValidation(fields, state, formValue);

    const [ newValidation, newErrorMessage ] = unzip(validationResults);

    state.setValidation(newValidation);

    state.setErrorMessage(newErrorMessage);

    return allEqual<boolean>(true, newValidation);
}

const formValue = <T extends unknown>(values: Value[], fieldIds: T[]) => (id: T): Value => {
    const index = fieldIds.indexOf(id)

    if (index === -1 || typeof values[index] === "undefined")
        throw new Error(`Could not find form value with id ${id}`);

    return values[index];
}

export type PropsToInitialValuesFunc<T_TargetComponentProps, T_FieldType> = (props: T_TargetComponentProps) => Map<T_FieldType, Value>

export type Options<T_FieldType, T_TargetComponentProps> = {
    errorMessagesVisible    : boolean;
    propsToInitialValues?   : PropsToInitialValuesFunc<T_TargetComponentProps, T_FieldType>
}

const showFormErrorMessages = <T_FieldType, T_TargetComponentProps>(
    options       : Options<T_FieldType, T_TargetComponentProps>,
    setOptions    : (options: Options<T_FieldType, T_TargetComponentProps>) => void,
    visible       : boolean,
) => (): void => {
    setOptions({ ...options, errorMessagesVisible: visible });
}

const disableForm = (setFormActive: (active: boolean) => void) => () => 
    setFormActive(false)

const enableForm = (setFormActive: (active: boolean) => void) => () => 
    setFormActive(true)

const initialValuesForForm = <T_FieldType extends unknown, T_TargetComponentProps extends unknown>(
    fields         : Field<T_FieldType>[],
    formOptions    : Options<T_FieldType, T_TargetComponentProps>,
    props          : T_TargetComponentProps
) => (): Value[] => {
    if (!formOptions.propsToInitialValues)
        return fields.map(getFieldInitial(""))

    const initialValuesMap = formOptions.propsToInitialValues(props)

    return fields.map(
        (field: Field<T_FieldType>) => initialValuesMap.get(field.id)
    ) as any
}

const setFormValues = <T extends unknown>(
    fields       : Field<T>[],
    state        : State,
    setTime      : (time: Date) => void,
) => (values: Map<T, Value>): void => {
    const newValues =
        fields.map((field: Field<T>, i: number) => values.get(field.id) || state.values[i])

    state.setValues(newValues)

    setTime(new Date())
}

const isFormFocused = (window: Window, inputIds: string[]): boolean => {
    if (!window.document.activeElement)
        return false

    return inputIds.indexOf(window.document.activeElement.id) > -1
}

const setEnterHandler = (inputIds: string[]) => (window: Window, handler: VoidFunction) => {
    const handlerWrapped = (event: KeyboardEvent) => {
        if (event.key === "Enter" && isFormFocused(window, inputIds)) {
            handler();
        }
    }

    window.addEventListener("keydown", handlerWrapped, true)

    return () => window.removeEventListener("keydown", handlerWrapped, true)
}

const buildInputIds = (formName: string, length: number, index = 0): string[] => {
    if (length === index - 1)
        return []

    return [ [ formName, "field", index ].join("_") ].concat(buildInputIds(formName, length, index + 1))
}

const getInvalidFields = <T extends unknown>(fields: Field<T>[], validation: boolean[], i = 0) => (): T[] => {
    if (i > fields.length - 1)
        return []

    const valid = validation[i]

    const field = fields[i]

    return [ ...(valid ? [] : [ field.id ]) ].concat(getInvalidFields(fields, validation, i + 1)())
}

const getFieldId = <T extends unknown>({ id }: Field<T>): T => id

const getFieldInitial = <T extends unknown>(defaultValue: Value) => ({ initial }: Field<T>): Value =>
    typeof initial === "undefined" ? defaultValue : initial

type FormBase<T> = React.ComponentClass<T> | React.FunctionComponent<T>

export const withForm = <T extends unknown, TargetComponentProps extends unknown>(
    formName          : string,
    fieldCollection   : FieldCollection,
    buildFields       : FieldsDefinition<T, TargetComponentProps>,
    TargetComponent   : FormBase<TargetComponentProps>,
    optionsBase?      : Options<T, TargetComponentProps>,
): React.FunctionComponent<TargetComponentProps> => {
    const defaultFormOptions: Options<T, TargetComponentProps> = {
        errorMessagesVisible    : true,
    }

    return (props: TargetComponentProps): React.ReactElement => {
        const formOptions = optionsBase || defaultFormOptions

        const [ fields ] = useState(() => buildFields(props))
        const [ values, setValues ] = useState<Value[]>(initialValuesForForm(fields, formOptions, props));
        const [ fieldIds ] = useState<T[]>(() => fields.map(getFieldId));
        const [ validation, setValidation ] = useState<boolean[]>(() => fields.map(() => true));
        const [ dirty, setDirty ] = useState(false);
        const [ errorMessage, setErrorMessage ] = useState<string[]>(() => fields.map(() => ""));
        const [ options, setOptions ] = useState<Options<T, TargetComponentProps>>(formOptions);
        const [ formActive, setFormActive ] = useState(true);
        const [ inputIds ] = useState(() => buildInputIds(formName, fields.length));
        const [ valueSetManuallyTime, setValueSetManuallyTime ] = useState(() => new Date());

        useEffect(() => {
            validateForm(fields, state, formValue(values, fieldIds))();
        }, [ valueSetManuallyTime ])

        const state: State =
            { values, validation, dirty, setValues, setValidation, setDirty, errorMessage, setErrorMessage };

        const getFormValue = formValue<T>(values, fieldIds) as ValueFunc

        const fieldComponents = toFieldComponents(
            formName,
            fields,
            state,
            new Map(),
            fieldCollection,
            options,
            getFormValue,
            formActive,
            inputIds,
        );

        const formProps: Props<T> = {
            field                  : mapGetter<T, React.ReactElement>(fieldComponents, "Could not find field component."),
            dirty,
            validate               : validateForm(fields, state, getFormValue),
            valid                  : allEqual<boolean>(true, validation),
            value                  : getFormValue,
            showErrorMessages      : showFormErrorMessages(options, setOptions, true),
            hideErrorMessages      : showFormErrorMessages(options, setOptions, false),
            disable                : disableForm(setFormActive),
            enable                 : enableForm(setFormActive),
            setValues              : setFormValues<T>(fields, state, setValueSetManuallyTime),
            setEnterHandler        : setEnterHandler(inputIds),
            getInvalidFields       : getInvalidFields<T>(fields, validation)
        }

        return (
            <TargetComponent
                form={formProps}
                { ...props }
            />
        );
    }
};

export const rangeValidator = (from: number, to: number, errorMessage: string) => (value: Value): [ boolean, string ] => {
    const valueType = typeof value

    if (valueType === "boolean")
        return [ false, errorMessage ]

    const num = valueType === "number" ? value : parseInt(value as string, 10)

    const inRange = !(num < from || num > to)

    return [ inRange, errorMessage ]
}

export const validateAll =
    <T extends unknown>(validators: FieldValidator<T>[]) => (value: Value, formValue: (id: T) => Value): FieldValidationResult =>
{
    if (validators.length === 0)
        return [ true, "" ]

    const validator = validators[0]

    const [ result, errorMessage ] = validator(value, formValue)

    if (!result)
        return [ false, errorMessage ]

    return validateAll(validators.slice(1))(value, formValue as any)
}
