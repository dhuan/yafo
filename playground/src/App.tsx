import React from 'react';
import './App.css';
import { withForm, fieldCollection, FormFieldType, FormValue, FormField, FormProps, parseCheckboxFormValue } from './yafo/yafo.js';

enum PlaygroundForm { FirstName, LastName, Country, Gender, Hobbies }

const countryOptions = ["Choose a country", "Brazil", "Germany"]
const genderOptions = ["Male", "Female"]
const checkboxOptions = ["Soccer", "Movies", "Music", "Books"]

const mapIndexToValue = <T extends unknown>(list: T[], indexList: number[]): T[] =>
    indexList.map(index => list[index])

const formFields = (): FormField<PlaygroundForm>[] => [
    {
        id        : PlaygroundForm.FirstName,
        label     : "First name",
        type      : FormFieldType.TEXT,
        valid     : (text: FormValue) => [ /[a-zA-Z{3,10}]/.test(text as string), "Invalid first name!" ],
        initial   : "",
        disabled  : false,
    },
    {
        id        : PlaygroundForm.LastName,
        label     : "Last name",
        type      : FormFieldType.TEXT,
        valid     : (text: FormValue) => [ /[a-zA-Z{3,10}]/.test(text as string), "Invalid last name!" ],
        initial   : "",
        disabled  : false,
    },
    {
        id        : PlaygroundForm.Country,
        label     : "Country",
        type      : FormFieldType.SELECT,
        valid     : (chosenCountry: FormValue) => [ chosenCountry > 0, "Choose a country!" ],
        initial   : 0,
        disabled  : false,
        options   : { selectOptions: countryOptions }
    },
    {
        id        : PlaygroundForm.Gender,
        label     : "Gender",
        type      : FormFieldType.RADIO,
        valid     : (chosenGender: FormValue) => [ chosenGender > -1, "Please chose a gender." ],
        initial   : -1,
        disabled  : false,
        options   : { radioOptions: genderOptions }
    },
    {
        id        : PlaygroundForm.Hobbies,
        label     : "Hobbies",
        type      : FormFieldType.CHECKBOX,
        valid     : (chosenHobbies: FormValue) => [ parseCheckboxFormValue(chosenHobbies as string).length > 0, "Please choose at least one hobby." ],
        initial   : "",
        disabled  : false,
        options   : { checkboxOptions: checkboxOptions }
    },
]

const FormComponentBase = ({ form, callback }: { form: FormProps<PlaygroundForm>, callback: Function }) => {
    return (
        <div>
            <h1>Playground Form</h1>

            <div id="field-first-name">
                { form.fieldComponents.get(PlaygroundForm.FirstName) }
            </div>

            <div id="field-last-name">
                { form.fieldComponents.get(PlaygroundForm.LastName) }
            </div>

            <div id="field-country">
                { form.fieldComponents.get(PlaygroundForm.Country) }
            </div>

            <div id="field-gender">
                { form.fieldComponents.get(PlaygroundForm.Gender) }
            </div>

            <div id="field-hobbies">
                { form.fieldComponents.get(PlaygroundForm.Hobbies) }
            </div>

            <input
                data-testid="submit"
                type="button"
                value="Submit Form"
                onClick={() => {
                    form.showErrorMessages()

                    if (!form.valid)
                        return

                    callback({
                        firstName: form.value(PlaygroundForm.FirstName),
                        lastName: form.value(PlaygroundForm.LastName),
                        country: countryOptions[form.value(PlaygroundForm.Country) as number],
                        gender: genderOptions[form.value(PlaygroundForm.Gender) as number],
                        hobbies: mapIndexToValue(checkboxOptions, parseCheckboxFormValue(form.value(PlaygroundForm.Hobbies) as string)).join(","),
                    })
                }}
            />
        </div>
    );
}

const FormComponent: any = withForm(
    "playground_form",
    fieldCollection,
    formFields,
    FormComponentBase,
    { errorMessagesVisible: false }
)

function App() {
  const [ formValues, setFormValues ] = React.useState<any>({})

  const formHasBeenSubmitted = Object.keys(formValues).length > 0

  return (
    <div className="app">
        <div className="form">
            <FormComponent callback={setFormValues} />
        </div>
        <div className="result">
            <h3>Form Results:</h3>
            { formHasBeenSubmitted && (
                <div>
                    <div>First Name: { formValues.firstName }</div>
                    <div>Last Name: { formValues.lastName }</div>
                    <div>Country: { formValues.country }</div>
                    <div>Gender: { formValues.gender }</div>
                    <div>Hobbies: { formValues.hobbies }</div>
                </div>
            ) }

            { !formHasBeenSubmitted && (
                <span>
                    Once you submit the form, the results shall appear here.
                </span>
            ) }
        </div>
    </div>
  );
}

export default App;
