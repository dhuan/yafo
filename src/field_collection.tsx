import React from 'react';
import { FormFieldComponentProps } from "./yafo"
import { FORM_FIELD_TYPE } from "./types"
import { toNumber } from "./utils"

const styles = {
    fieldWrapper: {
        paddingBottom: 14,
    },
    errorMessage: {
        paddingTop: 8,
        color: "red",
    }
}

const text: React.StatelessComponent<FormFieldComponentProps> =
    ({ inputId, label, value, disabled, onChange, errorMessage, options }: FormFieldComponentProps) =>
{
    return (
        <div style={styles.fieldWrapper}>
            <div>
                { label }
            </div>

            <div>
                <input
                    id={inputId}
                    type={options.password ? "password" : "text"}
                    value={value}
                    onChange={(event: any) => onChange(event.target.value)}
                    disabled={disabled}
                />
            </div>

            { errorMessage !== "" && (
                <div style={styles.errorMessage}>
                    { errorMessage }
                </div>
            ) }
        </div>
    )
}

const select: React.StatelessComponent<FormFieldComponentProps> =
    ({ inputId, label, value, disabled, onChange, errorMessage, options }: FormFieldComponentProps) =>
{
    const selectProps = {
    }

    return (
        <div style={styles.fieldWrapper}>
            <div>
                { label }
            </div>

            <select
                id={inputId}
                onChange={(event: any) => {onChange(toNumber(event.target.value))}}
                value={value}
                disabled={disabled}
            >
                { (options.selectOptions as string[]).map((option, i) => (
                    <option
                        key={["select", label, i].join("-")}
                        value={i}
                    >
                        { option }
                    </option>
                )) }
            </select>

            { errorMessage !== "" && (
                <div style={styles.errorMessage}>
                    { errorMessage }
                </div>
            ) }
        </div>
    )
}

export const fieldCollection = new Map([
    [ FORM_FIELD_TYPE.TEXT, text ],
    [ FORM_FIELD_TYPE.SELECT, select ],
])
