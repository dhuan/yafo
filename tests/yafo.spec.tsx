import '@testing-library/jest-dom'
import React from "react"
import renderer from "react-test-renderer";
import {act} from "@testing-library/react"
import { renderTestForm, getTestComponent, TestForm, callValidator } from "./utils";
import { validate } from "../src/yafo"
import loginForm, { LoginForm } from "./forms/login"

test("renders", () => {
    const Component = getTestComponent() as any

    expect(renderer.create(<Component callback={jest.fn()} />)).toMatchSnapshot();
})

test("get values", async () => {
    const { click, spy, fillFormWithValidFields } = renderTestForm()

    fillFormWithValidFields()

    click("[data-testid=submit]")

    expect(spy()).toHaveBeenCalledWith({
        firstName: "My first name",
        lastName: "My last name",
        country: 2,
        gender: 1,
        hobbies: "1,2",
    })
})

describe("Form Properties", () => {
    test("form.dirty", async () => {
        const { form, change } = renderTestForm()

        expect(form().dirty).toEqual(false);

        change("#field-first-name input[type=text]", "My first name")

        expect(form().dirty).toEqual(true);
    })

    test("form.value", async () => {
        const { form, change } = renderTestForm()

        expect(form().value(TestForm.FirstName)).toEqual("");

        change("#field-first-name input[type=text]", "My first name")

        expect(form().value(TestForm.FirstName)).toEqual("My first name");
    })

    test("form.valid", async () => {
        const { form, fillFormWithValidFields } = renderTestForm()

        expect(form().valid).toEqual(false)

        fillFormWithValidFields()

        expect(form().valid).toEqual(true)
    })
})

describe("Form Methods", () => {
    describe("form.field", () => {
        test("retrieving component", async () => {
            const { form, container } = renderTestForm()

            expect(form().field(TestForm.FirstName)).toMatchSnapshot()
        })

        test("attempting to retrieve unexisting component", async () => {
            const { form, container } = renderTestForm()

            expect(() => form().field(TestForm.Unused)).toThrowError("Could not find field component")
        })
    })

    test("form.setValues", async () => {
        const { form, container } = renderTestForm()

        const FIRST_NAME_INPUT_ELEMENT_SELECTOR = "#test_form_field_0"
        const LAST_NAME_INPUT_ELEMENT_SELECTOR = "#test_form_field_1"

        expect(container.querySelector<HTMLInputElement>(FIRST_NAME_INPUT_ELEMENT_SELECTOR)!.value).toEqual("")

        expect(container.querySelector<HTMLInputElement>(LAST_NAME_INPUT_ELEMENT_SELECTOR)!.value).toEqual("")

        act(() => {
            form().setValues(new Map([
                [ TestForm.FirstName, "Some first name" ],
                [ TestForm.LastName, "Some last name" ],
            ]))
        })

        expect(container.querySelector<HTMLInputElement>(FIRST_NAME_INPUT_ELEMENT_SELECTOR)!.value).toEqual("Some first name")

        expect(container.querySelector<HTMLInputElement>(LAST_NAME_INPUT_ELEMENT_SELECTOR)!.value).toEqual("Some last name")
    })

    test("form.enable and form.disable", async () => {
        const FIRST_NAME_INPUT_ELEMENT_SELECTOR = "#test_form_field_0"

        const { form, container } = renderTestForm()

        expect(container.querySelector<HTMLInputElement>(FIRST_NAME_INPUT_ELEMENT_SELECTOR)!.disabled).toEqual(false)

        act(() => {
            form().disable()
        })

        expect(container.querySelector<HTMLInputElement>(FIRST_NAME_INPUT_ELEMENT_SELECTOR)!.disabled).toEqual(true)

        act(() => {
            form().enable()
        })

        expect(container.querySelector<HTMLInputElement>(FIRST_NAME_INPUT_ELEMENT_SELECTOR)!.disabled).toEqual(false)
    })

    test("form.showErrorMessages and form.hideErrorMessages", async () => {
        const { form, container } = renderTestForm()

        const FIRST_NAME_ERROR_ELEMENT_SELECTOR = "div#error_message_test_form_field_1"

        expect(container.querySelectorAll(FIRST_NAME_ERROR_ELEMENT_SELECTOR).length).toEqual(0)

        act(() => {
            form().showErrorMessages()
        })

        expect(container.querySelectorAll(FIRST_NAME_ERROR_ELEMENT_SELECTOR).length).toEqual(1)

        act(() => {
            form().hideErrorMessages()
        })

        expect(container.querySelectorAll(FIRST_NAME_ERROR_ELEMENT_SELECTOR).length).toEqual(0)
    })

    test("form.getInvalidFields", async () => {
        const { form, change } = renderTestForm()

        expect(form().getInvalidFields()).toEqual([
            TestForm.FirstName,
            TestForm.LastName,
            TestForm.Country,
            TestForm.Gender,
            TestForm.Hobbies,
        ])

        change("#field-first-name input[type=text]", "Some invalid name 123")

        expect(form().getInvalidFields().indexOf(TestForm.FirstName)).toBeGreaterThan(-1)

        change("#field-first-name input[type=text]", "First Name")

        expect(form().getInvalidFields().indexOf(TestForm.FirstName)).toEqual(-1)
    })
})

describe("Utilities", () => {
    describe("validate", () => {
        test("validate.regex", () => {
            const validator = validate.regex(/^[a-z]{6}$/, "Invalid!")

            const callTestValidator = callValidator(validator)

            expect(callTestValidator("foobar")).toEqual([ true, "" ])

            expect(callTestValidator("foobar!")).toEqual([ false, "Invalid!" ])
        })

        test("validate.minLength", () => {
            const validator = validate.minLength(5, "Invalid!")

            const callTestValidator = callValidator(validator)

            expect(callTestValidator("1234")).toEqual([ false, "Invalid!" ])

            expect(callTestValidator("12345")).toEqual([ true, "" ])

            expect(callTestValidator("123456")).toEqual([ true, "" ])
        })

        test("validate.maxLength", () => {
            const validator = validate.maxLength(5, "Invalid!")

            const callTestValidator = callValidator(validator)

            expect(callTestValidator("1234")).toEqual([ true, "" ])

            expect(callTestValidator("12345")).toEqual([ true, "" ])

            expect(callTestValidator("123456")).toEqual([ false, "Invalid!" ])

            expect(callTestValidator("1234567")).toEqual([ false, "Invalid!" ])
        })

        test("validate.equalsField", () => {
            const { form, change } = renderTestForm({
                formComponent: loginForm.formComponent,
                formFields: loginForm.formFields,
            })

            change("#field-user input[type=text]", "some_user")

            change("#field-password input[type=text]", "some_password")

            expect(form().getInvalidFields().indexOf(LoginForm.Password)).toEqual(-1)

            act(() => {
                form().validate()
            })

            expect(form().getInvalidFields().indexOf(LoginForm.PasswordRepeat)).toBeGreaterThan(-1)

            change("#field-password-repeat input[type=text]", "some_different_password")

            expect(form().getInvalidFields().indexOf(LoginForm.PasswordRepeat)).toBeGreaterThan(-1)

            change("#field-password-repeat input[type=text]", "some-password")

            expect(form().getInvalidFields().indexOf(LoginForm.PasswordRepeat)).toBeGreaterThan(-1)

            change("#field-password-repeat input[type=text]", "some_password")

            expect(form().getInvalidFields().indexOf(LoginForm.PasswordRepeat)).toEqual(-1)
        })

        test("validate.equals", () => {
            const validator = validate.equals("foobar", "Invalid!")

            const callTestValidator = callValidator(validator)

            expect(callTestValidator("123")).toEqual([ false, "Invalid!" ])

            expect(callTestValidator("hello")).toEqual([ false, "Invalid!" ])

            expect(callTestValidator("foobar")).toEqual([ true, "" ])
        })

        test("validate.checkbox.min", () => {
            // The hobbies checkbox field is only valid if 2 hobbies are selected

            const { form, clickNth } = renderTestForm()

            // No hobbies selected: invalid
            expect(form().getInvalidFields().indexOf(TestForm.Hobbies)).toBeGreaterThan(-1)

            clickNth("input[type=checkbox]", 0)

            // Only one hobby selected: invalid
            expect(form().getInvalidFields().indexOf(TestForm.Hobbies)).toBeGreaterThan(-1)

            clickNth("input[type=checkbox]", 1)

            // 2 hobbies selected: valid
            expect(form().getInvalidFields().indexOf(TestForm.Hobbies)).toEqual(-1)

            clickNth("input[type=checkbox]", 2)

            // 3 hobbies selected: valid
            expect(form().getInvalidFields().indexOf(TestForm.Hobbies)).toEqual(-1)
        })

        test("validate.all", () => {
            const validator = validate.all([
                validate.minLength(5, "Need more!"),
                validate.maxLength(7, "Need less!"),
            ])

            const callTestValidator = callValidator(validator)

            expect(callTestValidator("123")).toEqual([ false, "Need more!" ])

            expect(callTestValidator("1234")).toEqual([ false, "Need more!" ])

            expect(callTestValidator("12345")).toEqual([ true, "" ])

            expect(callTestValidator("123456")).toEqual([ true, "" ])

            expect(callTestValidator("1234567")).toEqual([ true, "" ])

            expect(callTestValidator("12345678")).toEqual([ false, "Need less!" ])

            expect(callTestValidator("123456789")).toEqual([ false, "Need less!" ])
        })

        test("validate.none", () => {
            const validator = validate.none

            const callTestValidator = callValidator(validator)

            expect(callTestValidator("123")).toEqual([ true, "" ])

            expect(callTestValidator("1234")).toEqual([ true, "" ])
        })
    })
})
