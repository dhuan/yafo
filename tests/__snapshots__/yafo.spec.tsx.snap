// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Methods form.field retrieving component 1`] = `
<text
  custom={Object {}}
  disabled={false}
  errorMessage=""
  inputId="test_form_field_0"
  label="First name"
  onChange={[Function]}
  options={Array []}
  value=""
/>
`;

exports[`renders 1`] = `
<div>
  <h1>
    My Form
  </h1>
  <div
    id="field-first-name"
  >
    <div
      style={
        Object {
          "paddingBottom": 14,
        }
      }
    >
      <div>
        First name
      </div>
      <div>
        <input
          disabled={false}
          id="test_form_field_0"
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
  <div
    id="field-last-name"
  >
    <div
      style={
        Object {
          "paddingBottom": 14,
        }
      }
    >
      <div>
        First name
      </div>
      <div>
        <input
          disabled={false}
          id="test_form_field_1"
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
  <div
    id="field-country"
  >
    <div
      style={
        Object {
          "paddingBottom": 14,
        }
      }
    >
      <div>
        Country
      </div>
      <select
        disabled={false}
        id="test_form_field_2"
        onChange={[Function]}
        value={0}
      >
        <option
          value={0}
        >
          Choose a country
        </option>
        <option
          value={1}
        >
          Brazil
        </option>
        <option
          value={2}
        >
          Germany
        </option>
      </select>
    </div>
  </div>
  <div
    id="field-gender"
  >
    <div
      style={
        Object {
          "paddingBottom": 14,
        }
      }
    >
      <div>
        Gender
      </div>
      <div
        style={
          Object {
            "opacity": "1",
          }
        }
      >
        <div
          onClick={[Function]}
        >
          <input
            checked={false}
            name="test_form_field_3"
            readOnly={true}
            type="radio"
            value={0}
          />
          Male
        </div>
        <div
          onClick={[Function]}
        >
          <input
            checked={false}
            name="test_form_field_3"
            readOnly={true}
            type="radio"
            value={1}
          />
          Female
        </div>
      </div>
    </div>
  </div>
  <div
    id="field-hobbies"
  >
    <div
      style={
        Object {
          "paddingBottom": 14,
        }
      }
    >
      <div>
        Hobbies
      </div>
      <div
        style={
          Object {
            "opacity": "1",
          }
        }
      >
        <div
          onClick={[Function]}
        >
          <input
            checked={false}
            name="test_form_field_4"
            readOnly={true}
            type="checkbox"
            value={0}
          />
          Soccer
        </div>
        <div
          onClick={[Function]}
        >
          <input
            checked={false}
            name="test_form_field_4"
            readOnly={true}
            type="checkbox"
            value={1}
          />
          Movies
        </div>
        <div
          onClick={[Function]}
        >
          <input
            checked={false}
            name="test_form_field_4"
            readOnly={true}
            type="checkbox"
            value={2}
          />
          Music
        </div>
        <div
          onClick={[Function]}
        >
          <input
            checked={false}
            name="test_form_field_4"
            readOnly={true}
            type="checkbox"
            value={3}
          />
          Books
        </div>
      </div>
    </div>
  </div>
  <input
    data-testid="submit"
    onClick={[Function]}
    type="button"
    value="Click me"
  />
</div>
`;
